org.eclipse.transformer/src/main/java/org/eclipse/transformer/action/impl/ZipActionImpl.java

```diff
@@ -257,13 +260,25 @@ private void applyZipStream(
 		String prevName = null;
 		String inputName = null;
 
+		final Map<String, Pair<ZipEntry, ByteData>> signatureFilesMap = new HashMap<>();
+
 		try {
 			for ( ZipEntry inputEntry;
 				  (inputEntry = zipInputStream.getNextEntry()) != null;
 				  prevName = inputName, inputName = null ) {
 
 				try {
 					inputName = FileUtils.sanitize(inputEntry.getName()); // Avoid ZipSlip
+					if (ElementAction.SIGNATURE_FILE_PATTERN.matcher(inputName).matches()) {
+						/*
+						 * Signature file resource: Store its contents in memory until we have finished processing
+						 * the input zip file. Only at that point will we know if any of its resources were modified,
+						 * in which case the signature files will have been invalidated and must be discarded.
+						 * Otherwise, they must be added back to the output zip file.
+						 */
+						signatureFilesMap.put(inputName, Pair.newInstance(inputEntry, collect(inputName, zipInputStream)));
+						continue;
```

Comment:
This design basically sets the signature files to the side and then writes them out at the end of the file if there are no modifications. However JarInputStream needs the signature files at the front of the jar. See https://github.com/bndtools/bnd/blob/a58a37f0178551debf0c5cefff4278e1effee20b/biz.aQute.bndlib/src/aQute/bnd/osgi/Jar.java#L584.

So I think you will need a two pass method where after processing the jar (and this should only be done for type != ActionType.ZIP), if the jar was changed and the jar had signatures, then you will need to strip out the signatures (pass 2).